### PR TITLE
Adds ATen fallback for scaled_dot_product_attention

### PR DESCRIPTION
**Pull request by prathikr**
### Description

Introduces an ATen fallback for `torch.nn.functional.scaled_dot_product_attention`. This operator was introduced in torch 2.0 and, since then, has had many updates including the implementation of memory efficient attention for V100 machines. The current torchscript exporter exports a subgraph for attention which does not provide the same memory savings that PyTorch's memory efficient attention kernel provides. Allowing fallback to PyTorch ATen op for attention helps mitigate memory spike issues for models leveraging memory efficient attention.

### Motivation and Context

Memory issues arose when integrating ONNX Runtime Training with AML Stable Diffusion.

### CODE REVIEW

File: orttraining/orttraining/python/training/ortmodule/_custom_gradient_registry.py [hunk 1]
```diff
@@ -276,3 +277,45 @@ def upsample_nearest3d_gradient():
 @register_gradient("org.pytorch.aten", "ATen", "upsample_bicubic2d", "vec")
 def upsample_bicubic2d_gradient():
     return _upsample_gradient("upsample_bicubic2d_backward", 2)
+
+
+ATEN_SDPA_FALLBACK = os.getenv("ORTMODULE_ATEN_SDPA_FALLBACK", None)
```
Comment (pengwa) [inline orttraining/orttraining/python/training/ortmodule/_custom_gradient_registry.py:282]: this will be tricky when some of them attention is masked, and the others are not masked, we cannot differentiate them.,

File: orttraining/orttraining/python/training/ortmodule/_custom_op_symbolic_registry.py [hunk 2]
```diff
@@ -969,3 +970,26 @@ def softmax(g, input, dim, dtype=None):
     softmax = g.op("Softmax", casted_input, axis_i=dim)
 
     return softmax
+
+
+ATEN_SDPA_FALLBACK = os.getenv("ORTMODULE_ATEN_SDPA_FALLBACK", None)
+if ATEN_SDPA_FALLBACK:
+    # based on the following internal PyTorch kernel for efficient attention:
+    # https://github.com/pytorch/pytorch/blob/c12a4f2e65ad41b739aab1a261e2336b4a79fcfb/aten/src/ATen/native/native_functions.yaml#L14778
+    @register_symbolic("scaled_dot_product_attention")
```
Comment (pengwa) [inline orttraining/orttraining/python/training/ortmodule/_custom_op_symbolic_registry.py:979]: if the backward op we override is  "_scaled_dot_product_efficient_attention", I think the correct foward op we should override is "_scaled_dot_product_efficient_attention", otherwise, the fw and bw might NOT be the right pair.

File: orttraining/orttraining/python/training/ortmodule/_custom_gradient_registry.py (hunk 1 above)
Comment (prathikr) [reply orttraining/orttraining/python/training/ortmodule/_custom_gradient_registry.py:282]: as per Vincent's suggestion, I am removing support for masked attention and focusing this solution only on unmasked which is what stable diffusion utilizes.

File: orttraining/orttraining/python/training/ortmodule/_custom_op_symbolic_registry.py (hunk 2 above)
Comment (prathikr) [reply orttraining/orttraining/python/training/ortmodule/_custom_op_symbolic_registry.py:979]: The nomenclature used here matches that used in [symbolic_opset14.py](https://github.com/pytorch/pytorch/blob/ef0511245a92bae7057c195dcae2efc237b96f16/torch/onnx/symbolic_opset14.py#L140). For this reason, I believe scaled_dot_product_attention is correct.

Comment (pengwa) [reply orttraining/orttraining/python/training/ortmodule/_custom_op_symbolic_registry.py:979]: I see. If so, I think the condition we can enable the fallback is: the selected kernel is _scaled_dot_product_efficient_attention not others, otherwise the fw and bw pair is not correct. Could you pls add some notes in the env var description. I don't know how to check the selected kernel though, maybe you have some suggestions for users to check.

File: orttraining/orttraining/python/training/ortmodule/_custom_gradient_registry.py (hunk 1 above)
Comment (pengwa) [reply orttraining/orttraining/python/training/ortmodule/_custom_gradient_registry.py:282]: if so, we'd better add an assert attn_mask is None in  
```def scaled_dot_product_attention(g, query, key, value, attn_mask=None, dropout_p=0.0, is_causal=False, scale=None):```

if assertion fail, then we are aware of it clearly. And shall we add some notes we only support attn_mask is None?

File: orttraining/orttraining/python/training/ortmodule/_custom_op_symbolic_registry.py (hunk 2 above)
Comment (prathikr) [reply orttraining/orttraining/python/training/ortmodule/_custom_op_symbolic_registry.py:979]: Added some more docs.

**Review by pengwa — APPROVED** [review]
